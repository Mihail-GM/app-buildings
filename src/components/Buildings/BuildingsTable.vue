<template>
	<div id="buildings-table-container" class="mb-16">
		<div id="buildings-table-loading" v-if="actionState === 'LOADING'">
			<v-progress-linear
				indeterminate
				color="green"
			>

			</v-progress-linear>
			<div class="d-flex justify-center align-center pa-10">
				Loading... Please wait
			</div>
		</div>

		<v-alert
			v-else-if="actionState === 'ERRORED'"
			id="buildings-table-errored"
			dense
			outlined
			type="error"
		>
			There was an error when loading data. Please try again later.
		</v-alert>

		<v-data-table
			v-else
			id="buildings-table"
			:headers="headers"
			:items="buildings"
			sort-by="primary_contact_email"
			class="elevation-2 px-10 text-xl-h4"
		>
			<template v-slot:item.img="{ item }">
				<div class="d-flex justify-center align-center py-1">
					<v-img
						v-if="item.type"
						class="pt-5 mt-2"
						max-height="60"
						max-width="60"
						:src="getImgUrl(item.type)"
						:alt="imageData[item.type].alt"
						:title="imageData[item.type].title"
					/>
				</div>
			</template>

			<template v-slot:top>
				<v-toolbar
					flat
				>
					<v-spacer></v-spacer>

					<BuildingCreateEditDialog
						:edited-item-prop.sync="editedItem"
						:key="dialogKeyCreateEditDialog + '-create-edit'"
						:edited-index-prop="editedIndex"
						v-model="dialog"
						@resetEditedItem="resetEditedItem"
						@save="save"
					/>

					<BuildingDeleteDialog
						:key="dialogKeyDeleteDialog + '-delete'"
						v-model="dialogDelete"
						:edited-item-prop.sync="editedItem"
						@deleteItemConfirm="deleteItemConfirm"
					/>
				</v-toolbar>
			</template>

			<template v-slot:item.actions="{ item, index }">
				<v-icon
					:id="'edit-building-button-' + index"
					class="mr-2"
					@click.stop="editItem(item)"
				>
					mdi-pencil
				</v-icon>

				<v-icon
					:id="'delete-building-button-' + index"
					@click.stop="openDeleteItemDialog(item)"
				>
					mdi-delete
				</v-icon>
			</template>
		</v-data-table>
	</div>
</template>

<script>

	import BuildingCreateEditDialog from "./BuildingCreateEditDialog";
	import BuildingDeleteDialog from "./BuildingDeleteDialog";
	import BuildingsService from "@/services/buildings.service";
	import {mapGetters} from "vuex";

	export default {
		name: 'BuildingsTable',

		components: {
			BuildingDeleteDialog,
			BuildingCreateEditDialog
		},

		data() {
			return {
				dialog: false,
				dialogDelete: false,
				dialogKeyCreateEditDialog: 0,
				dialogKeyDeleteDialog: 0,

				headers: [
					{text: 'ID', align: 'start', sortable: true, value: 'id'},
					{text: 'Name', align: 'start', sortable: true, value: 'name'},
					{text: 'Area', align: 'start', sortable: true, value: 'area'},
					{text: 'Location', align: 'start', sortable: true, value: 'location'},
					{text: 'Type', align: 'start', sortable: true, value: `type`},
					{text: 'Image', align: 'center', sortable: false, value: `img`},
					{text: '', value: 'actions', sortable: false, align: 'end'},
				],

				editedIndex: -1,
				defaultItem: {
					uuid: null,
					id: null,
					name: '',
					area: '',
					location: '',
					type: '',
				},
				editedItem: {
					uuid: null,
					id: null,
					name: '',
					area: '',
					location: '',
					type: '',
				},

				imageData: {
					residential: {
						path: 'img/type-apartment.png',
						alt: 'Type - residential',
						title: 'Residential'
					},

					industrial: {
						path: 'img/industrial-building.png',
						alt: 'Type - industrial',
						title: 'Industrial'
					},

					educational: {
						path: 'img/educational-building.png',
						alt: 'Type - educational',
						title: 'Educational'
					},

					special: {
						path: 'img/special-building.png',
						alt: 'Type - special',
						title: 'Special'
					},
				}
			}
		},

		methods: {
			getAllBuildings() {
				this.$store.dispatch('buildings/fetchBuildings');
			},

			editItem(item) {
				this.editedIndex = this.buildings.indexOf(item);
				this.dialogKeyCreateEditDialog++;
				this.editedItem = Object.assign({}, item);
				this.dialog = true;
			},

			openDeleteItemDialog(item) {
				this.editedIndex = this.buildings.indexOf(item);
				this.dialogKeyDeleteDialog++;
				this.editedItem = Object.assign({}, item);
				this.dialogDelete = true;
			},

			resetEditedItem() {
				this.editedIndex = -1;
				this.editedItem = Object.assign({}, this.defaultItem);
			},

			save(editedItemFromForm) {
				if (this.editedIndex > -1) {
					BuildingsService.editBuilding(editedItemFromForm)
						.then((res) => {
							this.$store.dispatch('buildings/fetchBuildings');

							this.$store.dispatch('snackbar/showSnackbarMessage', {
								message: 'Building was edited successfully',
								duration: 4000,
								mode: 'success'
							});
						});

					return;
				}

				BuildingsService.createBuilding(JSON.stringify(editedItemFromForm))
					.then((res) => {
						this.$store.dispatch('buildings/fetchBuildings');

						this.$store.dispatch('snackbar/showSnackbarMessage', {
							message: 'Building was added successfully',
							duration: 4000,
							mode: 'success'
						});

						return res;
					});
			},

			closeDelete() {
				this.dialogDelete = false;
				this.$nextTick(() => {
					this.resetEditedItem();
				});
			},

			deleteItemConfirm(deletedItem) {
				BuildingsService.deleteBuilding(deletedItem)
					.then((res) => {
						this.$store.dispatch('buildings/fetchBuildings');

						this.$store.dispatch('snackbar/showSnackbarMessage', {
							message: 'Building was deleted successfully',
							duration: 4000,
							mode: 'success'
						});

						return res;
					});

				this.closeDelete();
			},

			getImgUrl(type) {
				return require('@/assets/' + (this.imageData[type].path))
			}
		},

		computed: {
			...mapGetters(
				{
					buildings: 'buildings/getBuildings',
					actionState: 'buildings/getCurrentState'
				}
			)
		},
		created() {
			this.getAllBuildings();
		},
	}
</script>
