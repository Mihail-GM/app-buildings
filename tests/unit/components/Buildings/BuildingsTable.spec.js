import Vuetify from 'vuetify'
import Vuex from "vuex";

import {createLocalVue, mount} from '@vue/test-utils'
import {merge} from "lodash";

import BuildingsTable from "@/components/Buildings/BuildingsTable";
import buildingsStore from "../../../../src/store/modules/buildings";
import flushPromises from "flush-promises";


describe('BuildingsTable.vue', () => {
	const localVue = createLocalVue();
	localVue.use(Vuex);
	let vuetify;
	let getters;

	beforeEach(() => {
		vuetify = new Vuetify();
	})

	function createStore(overrides) {
		let currentState = "DATA_LOADED";
		let buildings = [
			{
				area: "55",
				id: "2",
				img: "",
				location: "Sofia",
				name: "Luxury house 3",
				type: "residential",
				uuid: "-MukqrOCyJiAZW1BCwI0",
			},
			{
				area: "55",
				id: "2",
				img: "",
				location: "Sofia",
				name: "Luxury house 3",
				type: "educational",
				uuid: "-MukqrOCyJiAZW1BCwI0",
			}
		];

		getters = {
			getBuildings: () => currentState,
			getCurrentState: () => buildings
		}

		const defaultStoreConfig = {
			modules: {
				buildings: {
					namespaced: true,
					state: {
						currentState,
						buildings,
					},
					getters: buildingsStore.getters,

				}
			}
		};

		return new Vuex.Store(merge(defaultStoreConfig, overrides));
	}

	const createWrapper = (overrides) => {
		const defaultMountingOptions = {
			localVue,
			vuetify,
			store: createStore(),

			mocks: {
				// Vue Router
				$router: {
					push: () => {
					}
				},
			},
		};

		return mount(BuildingsTable, merge(defaultMountingOptions, overrides))
	}

	it('should render component ', () => {
		const wrapper = createWrapper();

		expect(wrapper.is(BuildingsTable)).toBe(true);
	})

	it('should set state to DATA_LOADED ', async () => {
		const store = createStore()
		const wrapper = createWrapper({store});

		expect(store.state.buildings.currentState).toBe('DATA_LOADED')
	})

	it('should set buildings from store to computed property buildings ', async () => {
		const store = createStore()
		const wrapper = createWrapper({store});

		expect(wrapper.vm.buildings.length).toBe(2);
	})

	it('should open edit dialog when edit button is clicked', async () => {
		const wrapper = createWrapper();

		let editButton = wrapper.find('#edit-building-button-0');
		editButton.trigger('click');
		await flushPromises();

		expect(wrapper.vm.dialog).toBe(true);
	})

	it('should open delete dialog when delete button is clicked', async () => {
		const wrapper = createWrapper();

		let editButton = wrapper.find('#delete-building-button-0');
		editButton.trigger('click');

		await flushPromises();

		expect(wrapper.vm.dialogDelete).toBe(true);
	})

	it('should show edited successfully message when call save with correct edit data', async () => {
		const store = createStore();
		store.dispatch = jest.fn(() => Promise.resolve());
		const wrapper = createWrapper({store});

		wrapper.setData({editedIndex: 2});
		let editedItem = {
			id: 5,
			name: 'Luxury building',
			area: '120',
			location: 'Str. Address',
			type: 'residential',
			img: '',
		}
		wrapper.vm.save(editedItem);

		await flushPromises();

		expect(store.dispatch).toHaveBeenCalledWith("snackbar/showSnackbarMessage", {
				message: 'Building was edited successfully',
				duration: 4000,
				mode: 'success'
			}
		);
	})

	it('should show add successfully message when call save with correct add data', async () => {
		const store = createStore();
		store.dispatch = jest.fn(() => Promise.resolve());
		const wrapper = createWrapper({store});

		wrapper.setData({editedIndex: -1});
		let editedItem = {
			id: 5,
			name: 'Luxury building',
			area: '120',
			location: 'Str. Address',
			type: 'residential',
			img: '',
		}
		wrapper.vm.save(editedItem);

		await flushPromises();

		expect(store.dispatch).toHaveBeenCalledWith("snackbar/showSnackbarMessage", {
				message: 'Building was added successfully',
				duration: 4000,
				mode: 'success'
			}
		);
	})

	it('should set editIndex to -1 when reseting data for dialogs', async () => {
		const wrapper = createWrapper();

		wrapper.setData({editedIndex: 5});
		let defaultItem = {
			id: null,
			name: '',
			area: null,
			location: '',
			type: null,
			img: '',
		};

		wrapper.vm.resetEditedItem(defaultItem);

		await flushPromises();

		expect(wrapper.vm.editedIndex).toBe(-1);
	})

	it('should set dialog to false when call closeDelete', async () => {
		const wrapper = createWrapper();

		await wrapper.setData({dialogDelete: true});
		await flushPromises();

		wrapper.vm.closeDelete();

		expect(wrapper.vm.dialogDelete).toBe(false);
		expect(wrapper.vm.editedIndex).toBe(-1);
	})

	it('should match snapshot', () => {
		const wrapper = createWrapper()

		expect(wrapper.html()).toMatchSnapshot()
	})
})
