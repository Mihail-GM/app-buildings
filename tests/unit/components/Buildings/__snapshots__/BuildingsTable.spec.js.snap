// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuildingsTable.vue should match snapshot 1`] = `
<div id="buildings-table-container" class="mb-16">
  <div class="v-data-table elevation-2 px-10 text-xl-h4 v-data-table--has-top v-data-table--has-bottom theme--light" id="buildings-table">
    <header class="v-sheet theme--light v-toolbar v-toolbar--flat" style="height: 64px;">
      <div class="v-toolbar__content" style="height: 64px;">
        <div class="spacer"></div>
        <div class="v-dialog__container"><button type="button" class="mb-2 px-8 v-btn v-btn--is-elevated v-btn--has-bg theme--dark v-size--default green" role="button" aria-haspopup="true" aria-expanded="false"><span class="v-btn__content">
			Add building
		</span></button>
          <!---->
        </div>
        <div class="v-dialog__container">
          <!---->
        </div>
      </div>
    </header>
    <div class="v-data-table__wrapper">
      <table>
        <colgroup>
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
        </colgroup>
        <thead class="v-data-table-header">
          <tr>
            <th role="columnheader" scope="col" aria-label="ID: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>ID</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Name: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Name</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Area: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Area</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Location: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Location</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Type: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Type</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Image" class="text-center"><span>Image</span></th>
            <th role="columnheader" scope="col" aria-label="" class="text-end"><span></span></th>
          </tr>
        </thead>
        <tbody>
          <tr class="">
            <td class="text-start">2</td>
            <td class="text-start">Luxury house 3</td>
            <td class="text-start">55</td>
            <td class="text-start">Sofia</td>
            <td class="text-start">residential</td>
            <td class="text-center">
              <div class="d-flex justify-center align-center py-1">
                <div aria-label="Type - residential" role="img" class="v-image v-responsive pt-5 mt-2 theme--light" style="max-height: 60px; max-width: 60px;" title="Residential">
                  <div class="v-responsive__content"></div>
                </div>
              </div>
            </td>
            <td class="text-end"><button type="button" id="edit-building-button-0" class="v-icon notranslate mr-2 v-icon--link mdi mdi-pencil theme--light"></button> <button type="button" id="delete-building-button-0" class="v-icon notranslate v-icon--link mdi mdi-delete theme--light"></button></td>
          </tr>
          <tr class="">
            <td class="text-start">2</td>
            <td class="text-start">Luxury house 3</td>
            <td class="text-start">55</td>
            <td class="text-start">Sofia</td>
            <td class="text-start">educational</td>
            <td class="text-center">
              <div class="d-flex justify-center align-center py-1">
                <div aria-label="Type - educational" role="img" class="v-image v-responsive pt-5 mt-2 theme--light" style="max-height: 60px; max-width: 60px;" title="Educational">
                  <div class="v-responsive__content"></div>
                </div>
              </div>
            </td>
            <td class="text-end"><button type="button" id="edit-building-button-1" class="v-icon notranslate mr-2 v-icon--link mdi mdi-pencil theme--light"></button> <button type="button" id="delete-building-button-1" class="v-icon notranslate v-icon--link mdi mdi-delete theme--light"></button></td>
          </tr>
        </tbody>
      </table>
    </div>
    <div class="v-data-footer">
      <div class="v-data-footer__select">Rows per page:<div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
          <div class="v-input__control">
            <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-443" class="v-input__slot">
              <div class="v-select__slot">
                <div class="v-select__selections">
                  <div class="v-select__selection v-select__selection--comma">10</div><input aria-label="Rows per page:" id="input-443" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
                </div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                </div><input type="hidden" value="10">
              </div>
              <div class="v-menu">
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="v-data-footer__pagination">1-2 of 2</div>
      <div class="v-data-footer__icons-before"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Previous page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></span></button></div>
      <div class="v-data-footer__icons-after"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Next page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></span></button></div>
    </div>
  </div>
</div>
`;
